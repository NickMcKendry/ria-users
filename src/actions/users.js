import database from '../firebase/firebase';

// ADD_USER

export const addUser = (user) => ({
  type: 'ADD_USER',
  user
});

export const startAddUser = (userData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      firstName = '',
      lastName = '',
      address = 0,
      createdAt = 0
    } = userData;

    const user = {
      firstName,
      lastName,
      address,
      createdAt
    };

    return database.ref(`users/${uid}/users`).push(user).then((ref) => {
      dispatch(addUser({
        id: ref.key,
        ...user
      }));
    });
  };
};

// REMOVE_USER

export const removeUser = ({ id } = {}) => ({
  type: 'REMOVE_USER',
  id
});

export const startRemoveUser = ({id} = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/users/${id}`).remove().then(() => {
      dispatch(removeUser({
        id
      }))
    })
  }
};

// EDIT_USER

export const editUser = (id, updates) => ({
  type: 'EDIT_USER',
  id,
  updates
});

export const startEditUser = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/users/${id}`).update(updates).then(() => {
      dispatch(editUser(id, updates))
    });
  };
};

// SET_USERS

export const setUsers = (users) => ({
  type: 'SET_USERS',
  users
});

export const startSetUsers = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/users`).once('value').then((snapshot) => {
      const users = [];

      snapshot.forEach((childSnapshot) => {
        users.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setUsers(users));
    });
  };
};
