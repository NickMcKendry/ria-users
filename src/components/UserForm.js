import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


export default class UserForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      firstName: props.user ? props.user.firstName : '',
      lastName: props.user ? props.user.lastName : '',
      address: props.user ? props.expense.address : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  };


  onFirstNameChange = (e) => {
    const firstName = e.target.value;
    this.setState(() => ({ firstName }));
  };

  onLastNameChange = (e) => {
    const lastName = e.target.value;
    this.setState(() => ({ lastName }));
  };

  onAddressChange = (e) => {
    const address = e.target.value;
    this.setState(() => ({ address }))
  };

  onDateChange = (createdAt) => {
    if(createdAt){
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  };

  onSubmit = ((e) => {
    e.preventDefault();

    if (!this.state.firstName || !this.state.lastName || !this.state.address) {
      this.setState(() => ({ error: 'Please provide a firstName, lastName and address!' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        firstName: this.state.firstName,
        lastName: this.state.lastName,
        createdAt: this.state.createdAt.valueOf(),
        address: this.state.address
      })
    }
  })

  render() {
    return (
        <form onSubmit={this.onSubmit} className="form">
          { this.state.error && <p className="form__error">{this.state.error}</p> }
          <input
            type="text"
            placeholder="First Name"
            autoFocus
            value={this.state.firstName}
            onChange={this.onFirstNameChange}
            className="text-input"
          />
          <input
            type="text"
            placeholder="Last Name"
            value={this.state.lastName}
            onChange={this.onLastNameChange}
            className="text-input"
          />
          <input
            type="text"
            placeholder="Address"
            value={this.state.address}
            onChange={this.onAddressChange}
            className="text-input"
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
            <div>
              <button className="button">Save User</button>
            </div>

        </form>
    )
  }
};
