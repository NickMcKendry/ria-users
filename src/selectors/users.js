import moment from 'moment';

export default (users, { text, sortBy, birthDate, endDate }) => {

  return users.filter((user) => {
    const createdAtMoment = moment(user.createdAt)

    const birthDateMatch = birthDate ? birthDate.isSameOrBefore(createdAtMoment, 'day') : true;

    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;

    const name = user.firstName + user.lastName

    const textMatch = name.toLowerCase().includes(text.toLowerCase());

    return birthDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if(sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if(sortBy === 'first-name') {
      return a.amount < b.amount ? 1 : -1;
    } else if(sortBy === 'last-name') {
      return a.amount < b.amount ? 1 : -1;
    } else if(sortBy === 'address') {
      return a.amount < b.amount ? 1 : -1;
    }
  })
}
