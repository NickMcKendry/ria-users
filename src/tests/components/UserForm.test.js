import React from 'react';
import { shallow } from 'enzyme';
import UserForm from '../../components/UserForm';
import users from '../fixtures/users';

test('Should render UserForm', () => {
  const wrapper = shallow(<UserForm />);
  expect(wrapper).toMatchSnapshot();
});

test('Should render UserForm with data', () => {
  const wrapper = shallow(<UserForm user={users[0]} />);
  expect(wrapper).toMatchSnapshot();
});

test('Should render error for invalid form submission', () => {
  const wrapper = shallow(<UserForm />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('Should set first name on input change', () => {
  const value = 'Nick';
  const wrapper = shallow(<UserForm />);
  wrapper.find('input').at(0).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('firstName')).toBe(value);
});

test('Should call onSubmit prop for valid form submit', () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(<UserForm user={users[0]} onSubmit={onSubmitSpy} />);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper.state('error')).toBe('');
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    firstName: users[0].firstName,
    lastName: users[0].lastName,
    address: users[0].address
  });
});
