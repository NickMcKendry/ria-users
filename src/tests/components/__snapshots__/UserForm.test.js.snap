// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render UserForm 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="First Name"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Last Name"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Address"
    type="text"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save User
    </button>
  </div>
</form>
`;

exports[`Should render UserForm with data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="First Name"
    type="text"
    value="Nick"
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Last Name"
    type="text"
    value="McKendry"
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Address"
    type="text"
    value="9803 w. 83rd ave."
  />
  <div>
    <button
      className="button"
    >
      Save User
    </button>
  </div>
</form>
`;

exports[`Should render error for invalid form submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="First Name"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Last Name"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Address"
    type="text"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save User
    </button>
  </div>
</form>
`;

exports[`Should render error for invalid form submission 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Please provide a firstName, lastName and address!
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="First Name"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Last Name"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Address"
    type="text"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save User
    </button>
  </div>
</form>
`;
